### PR TITLE
Closes #151

### LINKED ISSUE #151: Use `uint16` for recording coverage information
Ok, let's switch to `uint16` for coverage information then - almost no errors and half RAM usage for recording coverage information. Let's go for `uint8` only if RAM becomes a bottleneck.

_Originally posted by @leoisl in https://github.com/rmcolq/pandora/issues/148#issuecomment-506398232_

### PR DESCRIPTION
**Pull request by leoisl**
Closes https://github.com/rmcolq/pandora/issues/151

Switching to `uint16` for coverage information reduced the RAM usage for `pandora compare` on the full klebs dataset from 17.1 GB to 13.5 GB.

This dataset contains only 151 samples, and the more samples, the larger is the reduction, so this change is important for scalability.

### CODE REVIEW

File: include/kmergraphwithcoverage.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ class KmerGraphWithCoverage {
 private:
     //each coverage is represented as a std::vector, each position of the vector representing the coverage of a sample
     //each coverage of a sample is represented by a std::pair, for the forward and reverse reads coverage of that sample
-    std::vector<std::vector<std::pair<uint32_t, uint32_t>>> nodeIndex2SampleCoverage;
+    std::vector<std::vector<std::pair<uint16_t, uint16_t>>> nodeIndex2SampleCoverage;
```
Comment (mbhall88) [inline include/kmergraphwithcoverage.h]: snake_case

File: include/kmergraphwithcoverage.h [hunk 2]
```diff
@@ -47,23 +47,37 @@ class KmerGraphWithCoverage {
...
+    TODO: leaving return type as uint32_t to be safe for now, need a recheck later
+     */
     uint32_t get_covg(uint32_t node_id, bool strand, uint32_t sample_id) const;
     uint32_t get_num_reads() const { return num_reads; }
     uint32_t get_total_number_samples() const {return total_number_samples; }
 
     //setters
     void increment_covg(uint32_t node_id, bool strand, uint32_t sample_id);
-    void set_covg(uint32_t node_id, uint32_t value, bool strand, uint32_t sample_id);
+    void set_covg(uint32_t node_id, uint16_t value, bool strand, uint32_t sample_id);
```
Comment (mbhall88) [inline include/kmergraphwithcoverage.h]: change `strand` to `is_forward`

File: include/kmergraphwithcoverage.h [hunk 3]
```diff
@@ -47,23 +47,37 @@ class KmerGraphWithCoverage {
...
     void set_exp_depth_covg(const uint32_t);
     void set_binomial_parameter_p(const float);
     void set_negative_binomial_parameters(const float &, const float &);
     void set_thresh (int thresh) { this->thresh = thresh; }
     void set_num_reads(uint32_t num_reads) { this->num_reads = num_reads; }
 
     void zeroCoverages() {
-        for (auto &sampleCoverage: nodeIndex2SampleCoverage)
-            sampleCoverage = std::vector<std::pair<uint32_t, uint32_t>>(total_number_samples);
+        for (auto &sampleCoverage: nodeIndex2SampleCoverage) {
```
Comment (mbhall88) [inline include/kmergraphwithcoverage.h]: snake_case

**Comment by leoisl** [comment]
> As a more general statement. I would like if we could move away from having these enormous tests that test a huge number of different states and conditions. A test should test one single case and it should be fairly easy to see what the test is testing from reading the code.
> The test here that you have added realistically should be broken up into _many_ smaller tests. Additionally, the name of a test should tell the reader what the test does and what the expected outcome is. See [this test suite](https://github.com/rmcolq/pandora/blob/4ce991eee84dcc4439e80d89877cd036ab6904c5/test/denovo_discovery/denovo_discovery_test.cpp#L102-L298) for an example of testing a single function.

Indeed, you are right, it is a better approach. I will break this single test into 6 different tests.

File: include/kmergraphwithcoverage.h (hunk 1 above)
Comment (leoisl) [reply include/kmergraphwithcoverage.h]: Done in new commit.

File: include/kmergraphwithcoverage.h (hunk 2 above)
Comment (leoisl) [reply include/kmergraphwithcoverage.h]: Done in new commit.

File: include/kmergraphwithcoverage.h (hunk 3 above)
Comment (leoisl) [reply include/kmergraphwithcoverage.h]: Done in new commit.

**Comment by leoisl** [comment]
All your suggestions were done in new commit, please check if further improvements should be made.

**Comment by rmcolq** [comment]
I'm happy to merge/rebase this in now


-------- Original message --------

**Pull request merged**